pybrake/routes.py

```
@@ -0,0 +1,64 @@
+from threading import Lock, Timer
+from requests import post
+
+def route_stat_key(stat):
+  return "{}:{}:{}:{}".format(
+    stat["method"], stat["route"], stat["status_code"], stat["time"],
+  )
+
+class RouteStats():
+  def __init__(self, project_id, project_key, host):
+    self.stats = dict()
+    self._lock = Lock()
+
+    self._flush_thread = None
+    self._flush_period = 15.0
+    self._project_id = project_id
+    self._api_url = "{}/api/v4/projects/{}/routes-stats".format(host, project_id)
+    self._airbrake_headers = {
+      'Content-Type': 'application/json',
+      'Authorization': 'Bearer ' + project_key,
+    }
+
+  def inc_request(self, stat):
+    statKey = route_stat_key(stat)
+    with self._lock:
+      if statKey in self.stats:
+        stat = self.stats.get(statKey)
+      else:
+        stat["count"] = 0
+        stat["sum"] = 0
+        stat["sumsq"] = 0
+
+      stat["count"] = stat["count"] + 1
+      dur = stat["dur"]
+      stat["sum"] = stat["sum"] + dur
+      stat["sumsq"] += dur * dur
+
+      self.stats[statKey] = stat
+
+    self._init()
+
+  def _init(self):
+    if self._flush_thread is None or not self._flush_thread.is_alive():
+      self._flush_thread = Timer(self._flush_period, self._flush)
+      self._flush_thread.start()
+
+  def _flush(self):
+    stats = dict()
+    with self._lock:
+      stats = self.stats
+      self.stats = dict()
+
+    if not stats:
+      return
```

this should throw an error - stats always must be present

---

do what comment says